crates/fuel-gas-price-algorithm/gas-price-analysis/src/simulation.rs

```diff
@@ -1,162 +1,220 @@
-        .iter()
-        .zip(actual_reward_totals.iter())
-        .map(|(cost, reward)| *reward as i64 - *cost as i64)
-        .collect();
+            if let Some(mut da_blocks) = da_block.clone() {
+                let mut total_costs = updater.latest_known_total_da_cost;
+                for block in &mut da_blocks {
+                    total_costs += block.block_cost as u128;
+                    actual_costs.push(total_costs);
+                }
+                updater.update_da_record_data(da_blocks.to_owned()).unwrap();
+                assert_eq!(total_costs, updater.projected_total_da_cost);
+                assert_eq!(total_costs, updater.latest_known_total_da_cost);
+            }
+            updater
+                .update_l2_block_data(
+                    height,
+                    *fullness,
+                    capacity.try_into().unwrap(),
+                    *bytes,
+                    gas_price,
+                )
+                .unwrap();
+            da_gas_prices.push(updater.last_da_gas_price);
+            pessimistic_costs
+                .push(max_block_bytes as u128 * updater.latest_da_cost_per_byte);
+            actual_reward_totals.push(updater.total_da_rewards);
+            projected_cost_totals.push(updater.projected_total_da_cost);
+        }
 
-    let projected_profit: Vec<i64> = projected_cost_totals
-        .iter()
-        .zip(actual_reward_totals.iter())
-        .map(|(cost, reward)| *reward as i64 - *cost as i64)
-        .collect();
+        let (fullness_without_capacity, bytes): (Vec<_>, Vec<_>) =
+            fullness_and_bytes.iter().cloned().unzip();
+        let fullness = fullness_without_capacity
+            .iter()
+            .map(|&fullness| (fullness, capacity))
+            .collect();
+        let bytes_and_costs = bytes
+            .iter()
+            .zip(self.da_cost_per_byte.iter())
+            .map(|(bytes, cost_per_byte)| (*bytes, (*bytes * cost_per_byte) as u64))
+            .collect();
+
+        let actual_profit: Vec<i128> = actual_costs
+            .iter()
+            .zip(actual_reward_totals.iter())
+            .map(|(cost, reward)| *reward as i128 - *cost as i128)
+            .collect();
+
+        let projected_profit: Vec<i128> = projected_cost_totals
+            .iter()
+            .zip(actual_reward_totals.iter())
+            .map(|(cost, reward)| *reward as i128 - *cost as i128)
+            .collect();
+
+        SimulationResults {
+            gas_prices,
+            exec_gas_prices,
+            da_gas_prices,
+            fullness,
+            bytes_and_costs,
+            actual_profit,
+            projected_profit,
+            pessimistic_costs,
+        }
+    }
 
-    SimulationResults {
-        gas_prices,
-        exec_gas_prices,
-        da_gas_prices,
-        fullness,
-        bytes_and_costs,
-        actual_profit,
-        projected_profit,
-        pessimistic_costs,
+    fn zip_l2_blocks_with_da_blocks(
```

Comment:
A nit, but this could use some basic UT to see if it produces expected data.

Reply:
As a rule of thumb, I typically don't test internal/private functions. I understand that having some more confidence in this simulation code would be good though.

I don't think I want to hold this PR for this, but I'll think of ways I can test this all better.

Reply:
Makes sense, thanks for explanation.